modular_skyrat/modules/reagent_forging/code/forge.dm

```diff
@@ -133,556 +173,523 @@
 
-		if(!ceramic_item.forge_item)
-			to_chat(user, span_warning("You feel that setting [ceramic_item] would not yield anything useful!"))
-			return
+	var/obj/item/attacking_weapon = attacking_item
 
-		to_chat(user, span_notice("You start setting [ceramic_item]..."))
+	var/datum/component/reagent_weapon/weapon_component = attacking_weapon.GetComponent(/datum/component/reagent_weapon)
+	if(!weapon_component)
+		fail_message(user, "cannot imbue")
+		return
 
-		if(!do_after(user, ceramic_speed, target = src))
-			to_chat(user, span_warning("You stop setting [ceramic_item]!"))
-			return
+	if(length(weapon_component.imbued_reagent))
+		fail_message(user, "already imbued")
+		return
 
-		to_chat(user, span_notice("You finish setting [ceramic_item]..."))
-		var/obj/item/ceramic/spawned_ceramic = new ceramic_item.forge_item(get_turf(src))
-		user.mind.adjust_experience(/datum/skill/production, 50)
-		spawned_ceramic.color = ceramic_item.color
-		qdel(ceramic_item)
+	if(!do_after(user, 10 SECONDS, target = src))
+		fail_message(user, "stopped imbuing")
 		return
 
-	if(istype(attacking_item, /obj/item/stack/sheet/glass))
-		var/obj/item/stack/sheet/glass/glass_item = attacking_item
-		var/glassblowing_speed = user.mind.get_skill_modifier(/datum/skill/production, SKILL_SPEED_MODIFIER) * DEFAULT_TIMED
-		var/glassblowing_amount = DEFAULT_HEATED / user.mind.get_skill_modifier(/datum/skill/production, SKILL_SPEED_MODIFIER)
+	for(var/datum/reagent/weapon_reagent in attacking_weapon.reagents.reagent_list)
+		if(weapon_reagent.volume < MINIMUM_IMBUING_REAGENT_AMOUNT)
+			attacking_weapon.reagents.remove_all_type(weapon_reagent.type)
+			continue
 
-		if(in_use) //only insert one at a time
-			to_chat(user, span_warning("You cannot do multiple things at the same time!"))
-			return
-		in_use = TRUE
+		if(is_type_in_typecache(weapon_reagent, disallowed_reagents))
+			balloon_alert(user, "cannot imbue with [weapon_reagent.name]")
+			attacking_weapon.reagents.remove_all_type(weapon_reagent.type)
+			continue
 
-		if(forge_temperature < MIN_FORGE_TEMP)
-			fail_message(user, "The temperature is not hot enough to start heating [glass_item].")
-			return
+		weapon_component.imbued_reagent += weapon_reagent.type
+		attacking_weapon.name = "[weapon_reagent.name] [attacking_weapon.name]"
 
-		if(!glass_item.use(1))
-			fail_message(user, "You need to be able to use [glass_item]!")
-			return
+	attacking_weapon.color = mix_color_from_reagents(attacking_weapon.reagents.reagent_list)
+	balloon_alert_to_viewers("imbued [attacking_weapon]")
+	user.mind.adjust_experience(/datum/skill/smithing, 60)
+	playsound(src, 'sound/magic/demon_consume.ogg', 50, TRUE)
+	in_use = FALSE
+	return TRUE
 
-		if(!do_after(user, glassblowing_speed, target = src))
-			fail_message(user, "You stop heating up [glass_item]!")
-			return
+/// Handles clothing imbuing, extremely similar to weapon imbuing but not in the same proc because of how uhh... goofy the way this has to be done is
```

Comment:
I had two options, either make some REALLY ugly code, or split this into two procs and keep it kinda the same way it was before. Reagent forging is not long for this world once this and the crafting bench refactor get merged so I'm not sure if I want to spend time improving that.